From b720057301e891a3335379ac04ec9a10f054f029 Mon Sep 17 00:00:00 2001
From: ccoutant <ccoutant@138bc75d-0d04-0410-961f-82ee72b054a4>
Date: Thu, 15 May 2014 00:34:20 +0000
Subject: [PATCH] Change -g so that it will override -g1 but not -g3.

Backported from trunk r210442.

gcc/
        PR debug/61013
	* opts.c (common_handle_option): Don't special-case "-g".
	(set_debug_level): Default to at least level 2 with "-g".


git-svn-id: svn+ssh://gcc.gnu.org/svn/gcc/branches/gcc-4_9-branch@210456 138bc75d-0d04-0410-961f-82ee72b054a4
---
 gcc/ChangeLog |    6 ++++++
 gcc/opts.c    |   15 ++++++---------
 2 files changed, 12 insertions(+), 9 deletions(-)

diff --git a/gcc/ChangeLog b/gcc/ChangeLog
index 1ccda1c..9b4929b 100644
--- a/gcc/ChangeLog
+++ b/gcc/ChangeLog
@@ -1,3 +1,9 @@
+2014-05-14  Cary Coutant  <ccoutant@google.com>
+
+        PR debug/61013
+	* opts.c (common_handle_option): Don't special-case "-g".
+	(set_debug_level): Default to at least level 2 with "-g".
+
 2014-05-14  Eric Botcazou  <ebotcazou@adacore.com>
 
 	* config/sparc/sparc-protos.h (sparc_absnegfloat_split_legitimate):
diff --git a/gcc/opts.c b/gcc/opts.c
index faa63d6..1ce51bf 100644
--- a/gcc/opts.c
+++ b/gcc/opts.c
@@ -1820,13 +1820,8 @@ common_handle_option (struct gcc_options *opts,
       break;
 
     case OPT_g:
-      /* -g by itself should force -g2.  */
-      if (*arg == '\0')
-	set_debug_level (NO_DEBUG, DEFAULT_GDB_EXTENSIONS, "2", opts, opts_set,
-			 loc);
-      else
-	set_debug_level (NO_DEBUG, DEFAULT_GDB_EXTENSIONS, arg, opts, opts_set,
-			 loc);
+      set_debug_level (NO_DEBUG, DEFAULT_GDB_EXTENSIONS, arg, opts, opts_set,
+                       loc);
       break;
 
     case OPT_gcoff:
@@ -2076,10 +2071,12 @@ set_debug_level (enum debug_info_type type, int extended, const char *arg,
       opts_set->x_write_symbols = type;
     }
 
-  /* A debug flag without a level defaults to level 2.  */
+  /* A debug flag without a level defaults to level 2.
+     If off or at level 1, set it to level 2, but if already
+     at level 3, don't lower it.  */ 
   if (*arg == '\0')
     {
-      if (!opts->x_debug_info_level)
+      if (opts->x_debug_info_level < DINFO_LEVEL_NORMAL)
 	opts->x_debug_info_level = DINFO_LEVEL_NORMAL;
     }
   else
-- 
1.7.1

